### ISSUE TITLE
why does GSC and Bing Webmaster reject the sitemap?

### ISSUE THREAD

**Issue by jasonfb** [kjvarga/sitemap_generator#380]
For some reason GSC and Bing Webmaster tools are rejecting my sitemaps with no explanation.

the sitemap gets produced fine by the gem, it LOOKS ok to me (I've only been doing this professionally for 20 years but hey they switch around the rules every few years so it keeps me guessing), but then GSC and Bing both simply "do nothing" when I give them this site map.

it's infuriating actually, GSC literally does this:
![Sitemaps 2021-11-29 10-51-10](https://user-images.githubusercontent.com/59002/143900589-a9534cba-f808-4107-9d8e-dce6db1f27a3.png)

And Bing does this:
![Bing-no-response-bug](https://user-images.githubusercontent.com/59002/143900758-a426b0f9-a151-4302-89d8-2cf35c0ec0f4.gif)

Is this a sick joke? Do Google and Microsoft have professional programmers who work for them? No seriously, is this a sick joke? If any developer who worked for me wrote such a bad (non-informative) interface I would fire them.

I can't believe it. This worked for 25 years until this year. What is happening with software. 

please someone help. The sitemap is generated with this gem using:

```
SitemapGenerator::Sitemap.adapter = SitemapGenerator::AwsSdkAdapter.new( "#{ENV['S3_BUCKET_NAME']}",
                              aws_access_key_id: ENV["AWS_S3_ACCESS_KEY_ID"],
                              aws_secret_access_key: ENV["AWS_S3_SECRET_ACCESS_KEY"],
                              aws_region: ENV["S3_REGION"])
SitemapGenerator::Sitemap.ping_search_engines


SitemapGenerator::Sitemap.create(compress: false) do
...
```

**_It uploadsFINE to the  bucket, here:_**
https://helios-public-assets.s3.us-east-2.amazonaws.com/sitemap.xml

as far as I can tell, the gem is generating the sitemap just fine (or maybe I missed something?), the problem is Google and Bing have gone off the deep end and are refusing to accept sitemaps that worked fine since the history of the Web and won't give any error messages (outrageous)

(I also tried with  `.xml.gz` but those have the same problem so I switched it to `compress: false` but same effect)

I don't know if maybe something is on Amazon's end (I noticed they recently started redirecting requests to S3 that don't have the region name them to the canonical name with the region).... ?

What is happening here? Why won't GSC or Bing give me even a hint or an error message at what the problem is? Is this 2021 or 1991?

if this is the future it is already dystopia.

**Comment by jasonfb** [comment]
I have discovered that GSC no longer support sitemaps on foreign domains -- in theory they do-- but I cannot get mine setup.

so It seems that just loading it directly from the rails app will do, which I hope works on the ephemeral 12-Factor build.

**Comment by jrochkind** [comment]
Super annoying. I guess you could write a little Rails action to proxy it from S3 if you had to.

**Comment by jasonfb** [comment]
What's really frustrating here is that any kind of stuffing it into the app is a violation of the 12-factor deploy principles; because you'd have to make every instance of your app somehow be in sync.
The only other think I can think is to setup a proper route that somehow fetches it from a cache mechanism like Redis, which would be straightforward & fast.

**Comment by jasonfb** [comment]
here are the answers I got on stack exchange:


https://webmasters.stackexchange.com/questions/137706/google-search-console-rejects-a-valid-sitemap-with-no-explanation

and

https://webmasters.stackexchange.com/questions/137709/bing-webmaster-tools-does-nothing-when-i-try-to-upload-my-sitemap-spins-goe

https://www.sitemaps.org/protocol.html


Given this information, it seems to me like one should just abandon the idea of S3 uploading sitemaps to remote buckets.

Furthermore, the challenge with this gem then becomes that it is a violation of 12-factor deploy principles. With the S3 upload (previously), this worked with 12-factor, but this means it no longer is friendly to 12-factor

Does anyone know of any other alternatives?

Ideally the gem would do these things

• Generate site maps and store them in REDIS (strange I know)
• Create a fake route at `/sitemap.xml` that secretly hits a special controller, fetches the data from the Redis cache, and then returns it. 


this design would be friendly for a multi-dyno 12-factor setup

**Comment by jrochkind** [comment]
So I believe my sitemap is still working on S3, with a different hostname than the site it maps. But it was challenging and error-prone and I'm often not 100% sure I got it right. 

You don't need to store the sitemap in REdis to create a route (nothing fake about the route) at `/sitemap.xml` (or anywhere else; you can say where your sitemap resides in your robots.txt) that fetches the data and then returns it. It can fetch the data from S3 (or anywhere else) and then return it just as easily as it can fetch it from redis and then return it. 

The trick is that the sitemap isn't just one file, it can be several files if it gets big. Can/should be gzip'd etc. So it's more than just one URL, it has to handle/map all of them. 

Another trick is that _ideally_ for performance purposes, you don't want to fetch the entire sitemap file(s) (which can be quite large) into memory, and only once fetched then return it. you want to stream it, byte by byte, from the remote location (whether redis or S3) to the client. This is possible with Rails (the API kind of keeps changing from Rails version to Rails version though), but is another thing to figure out. 

It's true that this is kind of complicated to figure out and implement. It's also not technically anything wrong with this gem, this gem is doing what's advertised. You just kind of need that extra piece somehow if you want to have sitemap on the same hostname with "12 factor" (eg Heroku) deploy. That extra piece is kind of complementary. I am not sure if the maintainer of this gem is interested in developing it though. 

I am not a maintainer here, just another interested user.

**Comment by kjvarga** [comment]
👋  hi folks.  This discussion comes up quite a lot and people regularly explore complicated ways of making their sitemaps accessible from their applications.  But the truth is that it's not necessary.  It's possible to host sitemaps on a remote domain as long as you prove that you control that domain.  See https://github.com/kjvarga/sitemap_generator#upload-sitemaps-to-a-remote-host-using-adapters

The reason I don't try to support a Rails route to serve from the application is that as @jrochkind points out there are a lot of non-trivial things to consider.  And the fact is that everyone wants to do things slightly differently, with different data stores, different application scale, different needs.  S3 or another file server service is still the best and most efficient way to serve sitemaps, which are files.  There are no technical blockers to doing that that I am aware of, if you follow the correct steps.

**Comment by jrochkind** [comment]
I do see the potential use of a proxy route, although it's tricky to get right and figure out how to do with Rails... I may at some point try to write it as a third-party thing people can use, but no promises.

**Comment by jasonfb** [comment]
thank you all for the feedback!

I plan (at least have on my to-do list) a successor gem that will solve the 12-factor problem.

**Comment by jasonfb** [comment]
@kjvarga --- Ok afte review it looks like my frustration may be the GSC and/or Bing web interface (no failure message)

I cannot remember if I correctly did domain control on the S3 bucket-- which I can only do via an web method as of course I do not control the S3 domain ---  I will try again to make sure I do domain control first. 

as far I could tell, this was no longer possible and I suspect that GSC and/or Bing changed the rules but I will try it again.

**Comment by jasonfb** [comment]
@kjvarga -- I looked into it and what you have claimed above is now false information. Please provide evidence to back up your assertion "It's possible to host sitemaps on a remote domain as long as you prove that you control that domain"

I believe you may be familiar with grandfathered technology that could possibly be working on your websites but is now no longer possible. Like I said, i know for a fact this worked for about 15 years until last year.

In my attempts, I added both the S3 bucket as a "property" here:
![Overview 2022-01-11 10-46-12](https://user-images.githubusercontent.com/59002/148975312-4a180a5e-bd69-418b-9bbf-a4fc072cc3b1.png)

this is in addition to my real domain property.

When I go back to my real domain property and add the anything that is not part of the domain or subdomain on Sitemaps, I get this error:

![Sitemaps 2022-01-11 10-47-06](https://user-images.githubusercontent.com/59002/148975589-dd219bd4-975d-4243-b18d-95077ebd6282.png)

If I use any domain or subdomain of my website it accepts the sitemap URL (even if then telling me it couldn't fetch the Sitemap URL).

if I use ANY OTHER DOMAIN, it rejects it immediately, even if that domain is in the Properties.

You have outdated information — I am reopening this issue. Your claim that  "It's possible to host sitemaps on a remote domain as long as you prove that you control that domain" is **false information**.

**Comment by jasonfb** [comment]
@kjvarga -- I have some new information and good news. I am able to indeed get this working, but only by this setup:

1. create a Cloudfront distribution to the S3 bucket where the sitemap lives (so, by presumption one would use the S3 uploader option of this gem)
2. When you set up this Cloudfront distribution, you MUST set up an **Alternative Domain name**

Your alternative domain name must be a subdomain of your primary domain (assets1.yourdomain.com)

Then, AWS will make you set up a SSL cert for your alternative domain. I suggest using DNS verification here. In order to complete that step, you'll add some verification records to your domain https://docs.aws.amazon.com/acm/latest/userguide/dns-validation.html

You will be doing this all in a 2nd browser tab while in the 1st browser tab you will still be creating the CloudFront distribution. That's because you must validate the domain first before you can create the distribution with the alternative domain name on it

it takes several minutes of course for AWS to validate.

Then, once you successfully create the distro, you see this:
![CloudFront 2022-01-12 15-44-59](https://user-images.githubusercontent.com/59002/149219585-1e5e6778-9d17-49b3-acce-21eb3df811e5.png)

Notice that I have created the CF distribution telling it my alternative domain name is assets1.domain.com

However, I still will not be able to set the subdomain record for assets1.domain.com until after I create this CloudFront distribution-- that would be a Catch-2022-- because to create the DNS record I need its target.

3. Now with the Cloudfront distribution I can create a CNAME record for assets1 in my domain that points to this Cloudfront distro

4. Finally, back in GSC, you can now add the SSL-enabled asset subdomain.

![Sitemaps 2022-01-12 15-53-42](https://user-images.githubusercontent.com/59002/149220078-ea49a1ee-82c0-4488-af5c-787139f2ba88.png)


I like this solution as it keeps the ability to outport the sitemaps to S3, which is nice for 12-factor/distributed apps, and also to have the sitemaps "hot". 

However, based on this research it is now clear to me that prior to about 2019 or 2020, Google + Bing used to allow you to use the S3 buckets directly onto the sitemaps.

**If your websites still have GSC settings with S3 buckets, I suspect you are grandfathered into this feature. New websites do not appear to allow this in GSC.**

The reason this is significant is that the setup steps involved in correctly getting that TLS/SSL cert onto the asset subdomain are not insignificant, which really just means more work for us I guess. But I agree now that you are correct with the right subdomain setup— as explained above— it is still possible to use sitemaps on S3 via a CNAME

**Comment by jiggins89** [comment]
@jasonfb what are the sitemap config settings you used when doing the cloudfront option? I assume `SitemapGenerator::Sitemap.default_host` is still the domain rather than the subdomain? What do you use for `SitemapGenerator::Sitemap.sitemaps_host`??? Is it still the s3 bucket or the cloud distro?
